src/blog/articles/node-js-race-conditions/index.md

```diff
@@ -0,0 +1,351 @@
+
+  1. Read the current value for `balance`
+  2. Add `50` _aurei_ to it
+  3. Save the resulting value into `balance`
+
+Since the two components are running in parallel, without any synchronisation mechanism, the following case could happen:
+
+{% responsiveImage './blog/articles/node-js-race-conditions/aurei-race-condition-node-js.png', 'A race condition example showing 2 processes trying to update a balance', { maxWidth: 848 }  %}
+
+In the picture above you can see that **Component 2** ends up having a _stale_ view of the balance: the balance gets changed by **Component 1** after **Component 2** has read the balance. For this reason, when **Component 2** performs its own update, it is effectively overriding any change previously made by **Component 1**. This is why we have a race condition: the two components are effectively racing for completing their own tasks and they might end up stepping onto each other toes! This doesn't make [Julius](https://en.wikipedia.org/wiki/Julius_Caesar) happy I am afraid...
+
+One way to solve this problem is to isolate the 2 concurrent operations into _transactions_ and make sure that there is only one transaction running at a given time. This idea might look like this:
+
+{% responsiveImage './blog/articles/node-js-race-conditions/aurei-fixed-race-condition-node-js.png', 'Fixing a race condition using a transaction', { maxWidth: 848 }  %}
+
+In the last picture, we are using transactions to make sure that all the steps of **Component 1** happen in order before all the steps of **Component 2**. This avoids any _stale read_ and makes sure that every component always has an up to date view of the world before doing any change. You can stop holding your breath now, Julius!
+
+In the rest of this article, we will zoom in more on race conditions in the context of Node.js and we will see some other approaches that can allow dealing with them.
+
+
+## Can we have race conditions in Node.js?
+
+It is a common misconception to think that Node.js does not have race conditions because of its single-threaded nature. While it is true that in Node.js you would not have multiple threads competing for resources, you might still end up with events belonging to different logical transactions being executed in an order that might result in _stale reads_ and generate a race condition.
+
+In the example that we illustrated above, we intentionally represented the various events (_read_, _increase_ and _save_) as discrete units. Note how the system is never executing more than one event at the same time. This is a simple but accurate representation of how the Node.js event loop processes events on a single thread. Nonetheless, you can see that there might be situations where multiple logical transactions (e.g. multiple deposits) are scheduled concurrently on the event loop and the discrete events might end up being intermingled, which results in a race condition.
```

Comment:
I think it's important to specify that read, increase and save are not in-memory (nor synchronous) operations, which would be instantaneous in Node.js (e.g. saving to a file). Actually that's where Node.js single-thread comes handy. In a multi-threaded application, there would be the risk of race-conditions even if those were in-memory operations, while in Node.js those can never happen. Race conditions in Node can happen only when there is asynchronous code and the control is returned back to the event loop and another operation.